### ISSUE TITLE
Cloud Service Broker (0.2.7 / 1.0.0-rc.40) failed to subsume db with failover group with db storage size near the capacity

### ISSUE THREAD

**Issue by kelvin-j-li** [cloudfoundry-incubator/csb-brokerpak-azure#53]
**Describe the bug**
Cloud Service Broker (0.2.7 / 1.0.0-rc.40) failed to subsume db with failover group with db storage size near the capacity

**To Reproduce**
`manually inserting many data row:
-- Create Person table
CREATE TABLE Person
(
    PersonId INT IDENTITY PRIMARY KEY,
    FirstName NVARCHAR(1024) NOT NULL,
    LastName NVARCHAR(1024) NOT NULL
)
INSERT INTO Person VALUES('FFFFFFF', 'LLLLL');
DECLARE @cnt INT = 0;
WHILE @cnt < 146281
BEGIN
   INSERT INTO Person VALUES('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
   'LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL');
   SET @cnt = @cnt + 1;
END;
select count(*) from Person;
# count(*) - 146281`

**Expected behavior**
Please disable the destructive behaviour when csb subsume. In the above test, both the primary and secondary database were deleted from azure portal, as was the fail over group in the db server. 

**Screenshots**
![spring-music-db-full](https://user-images.githubusercontent.com/72498127/125604297-fd6ef022-0195-4e6d-88aa-991c515f5b7f.png)

**Additional context**
Use spring-music sample to to bind to the azure sqldb failover group instance.

**Comment by blgm** [comment]
Hi @kelvin-j-li,

Firstly, apologies for taking so long to start investigating this. Our aim is to be faster in responding to issues, and we hope to be faster in the future.

We tried to replicate this issue, but have not succeeded. Perhaps there is a difference in what we are doing, so I'll explain what we tried to do, and maybe you'll be able to see some differences.

We used slightly different versions: cloud-service-broker 0.2.8 and csb-brokerpak-azure 1.0.0-rc.41. But we don't think there have been any fixes for this issue, so it should not have made any difference.

## Attempt 1:
We attempted to subsume a database that was 100 MB in size (as in the description above), but we encountered a different error:

The steps followed were:

- Created a primary DB and a secondary one of the type basic and set the max storage size to 100MB.
- Filled the DB using the script above and that resulted in using 100% of all the DB space.
- Attempted to subsume it in CSB.

We observed the error:
```
{"timestamp":"1629307485.634397507","source":"Import","message":"Import.plan failed","log_level":2,"data":{"correlation-id":"4c7fbbbe-20e0-4d79-4cb1-ed8bc5f2e4df","error":"Error: expected max_size_gb to be in the range (1 - 4096), got 0  on main.tf line 2, in resource \"azurerm_mssql_database\" \"primary_db\":   2: resource \"azurerm_mssql_database\" \"primary_db\" { exit status 1","request-id":"b8d49fa9-f210-4848-baee-2c8f73440b9f"}}
```
This error is not an ideal behaviour (and we have raised ticket [#179317385](https://www.pivotaltracker.com/story/show/179317385) to investigate further), but it's not the same issue as above.

## Attempt 2:
```
# Create DB
cf create-service azure-sqldb basic azure-sqldb-basic-grass-spirit -c '{"resourceGroup":"ancient-myth","sqlServerName":"server-west-serpent","sqldbName":"db-jungle-venom"}'

# Create FOG
cf create-service azure-sqldb-failover-group SecondaryDatabaseWithFailoverGroup azure-sqldb-failover-group-SecondaryDatabaseWithFailoverGroup-rapid-crystal -c '{"failoverGroupName":"fog-flannel-carpet","primaryDbName":"db-jungle-venom","primaryServerName":"server-west-serpent","readWriteEndpoint":{"failoverPolicy":"Automatic","failoverWithDataLossGracePeriodMinutes":60},"secondaryServerName":"server-blossom-arm"}'

# Fill the database with data (we used an app to do this)

# Set up CSB with server details
cf set-env cloud-service-broker MSSQL_DB_FOG_SERVER_PAIR_CREDS '{"raven":{"admin_password":"********","admin_username":"sql-server-user","primary":{"resource_group":"ancient-myth","server_name":"server-west-serpent"},"secondary":{"resource_group":"ancient-myth","server_name":"server-blossom-arm"}}}'
cf set-env cloud-service-broker GSB_SERVICE_CSB_AZURE_MSSQL_DB_FAILOVER_GROUP_PROVISION_DEFAULTS '{"server_credential_pairs":{"raven":{"admin_password":"********","admin_username":"sql-server-user","primary":{"resource_group":"ancient-myth","server_name":"server-west-serpent"},"secondary":{"resource_group":"ancient-myth","server_name":"server-blossom-arm"}}}}'

# Subsume the database
cf create-service csb-azure-mssql-db-failover-group subsume csb-azure-mssql-db-failover-group-subsume-gem-bell -c '{"azure_fog_id":"/subscriptions/34595d5a-041e-11ec-b9fa-732dd74cc930/resourceGroups/ancient-myth/providers/Microsoft.Sql/servers/server-west-serpent/failoverGroups/fog-flannel-carpet","azure_primary_db_id":"/subscriptions/34595d5a-041e-11ec-b9fa-732dd74cc930/resourceGroups/ancient-myth/providers/Microsoft.Sql/servers/server-west-serpent/databases/db-jungle-venom","azure_secondary_db_id":"/subscriptions/34595d5a-041e-11ec-b9fa-732dd74cc930/resourceGroups/ancient-myth/providers/Microsoft.Sql/servers/server-blossom-arm/databases/db-jungle-venom","server_pair":"raven"}'
```
These commands were all successful. In the Azure change log, observed that some write database activities took place as the database was subsumed, and these related to tags, SecurityAlertPolicies, etc... But we did not observe either the primary, secondary or FOG being deleted.

- Do you see any differences between what we have done and what you did that might be significant?
- Is there anything else you can tell us about what occurred (e.g. some logs etc...) that might shed light on what is happening?
- Cloud Service Broker 0.2.8 has a change that runs `terraform plan` during a subsume to try and detect situations where a subsequent `terraform apply` would be destructive. If you have an opportunity, it would be interesting to know whether or not this makes any difference to this scenario. The expectation is that it would convert a destructive subsume into a failure with some additional information in the log.

**Comment by kelvin-j-li** [comment]
Hi George,
Thanks for the update.

The initial reactive destructive issues of csb subsume are the below:

![image](https://user-images.githubusercontent.com/72498127/130540198-cf7feb71-e438-4f97-a2c2-d838a9c29b9b.png)

The details are also posted in **Slack connection** <**cloudservicebroker**>
https://manulife-ets-dpse.slack.com/archives/C0164KKEZGX/p1629768307002200

When we tested with CSB 0.2.7, we reduced the db size from 2G to 100MB to make it easier to get to the situation of db storage is nearly. Apparently it seems to trigger a different use case here -- minimum db size is GB so any smaller size is round off to 0. 

I will deploy 0.2.8 with broker pak 41rc to re-test item 1 (with high DTU usage) and item 2 (100MB and 2G with 96% disk usage), to see whether FR 271 can prevent the destructive outcome of above item 1 and 2. 

Thank you!

**Comment by blgm** [comment]
@kelvin-j-li,

With a broker version earlier than 0.2.8 and a database that is smaller than 1GB in size, I am seeing a failure to subsume. For now I'm treating this as a [separate issue](https://www.pivotaltracker.com/story/show/179317385).
```
$ cf create-service csb-azure-mssql-db-failover-group subsume csb-azure-mssql-db-failover-group-subsume-puddle-drop --wait -c {"azure_fog_id":"/subscriptions/e0c2ac72-04be-11ec-8ae5-37eb8cfeb7c1/resourceGroups/cake-viper/providers/Microsoft.Sql/servers/server-ancient-face/failoverGroups/fog-shimmer-kitten","azure_primary_db_id":"/subscriptions/e0c2ac72-04be-11ec-8ae5-37eb8cfeb7c1/resourceGroups/cake-viper/providers/Microsoft.Sql/servers/server-ancient-face/databases/db-bone-watcher","azure_secondary_db_id":"/subscriptions/e0c2ac72-04be-11ec-8ae5-37eb8cfeb7c1/resourceGroups/cake-viper/providers/Microsoft.Sql/servers/server-fern-lifter/databases/db-bone-watcher","server_pair":"face"}

$ cf service csb-azure-mssql-db-failover-group-subsume-puddle-drop
Showing info of service csb-azure-mssql-db-failover-group-subsume-puddle-drop in org pivotal / space broker-cf-test as admin...

name:            csb-azure-mssql-db-failover-group-subsume-puddle-drop
guid:            2b85a699-94ee-4585-8c6c-c420adac6a94
type:            managed
broker:          broker-cf-test
offering:        csb-azure-mssql-db-failover-group
plan:            subsume
tags:
offering tags:   azure, mssql, sqlserver, dr, failover, preview
description:     Manages auto failover group db's on existing Azure SQL server pairs
documentation:   https://docs.microsoft.com/en-us/azure/sql-database/sql-database-auto-failover-group/
dashboard url:

Showing status of last operation:
   status:    create failed
   message:   Error: expected max_size_gb to be in the range (1 - 4096), got 0  on main.tf line 2, in resource "azurerm_mssql_database" "primary_db":   2: resource "azurerm_mssql_database" "primary_db" { exit status 1
   started:   2021-08-24T09:55:43Z
   updated:   2021-08-24T09:55:43Z
...

$ cf logs cloud-service-broker --recent
...
   2021-08-24T10:55:21.03+0100 [APP/PROC/WEB/0] ERR {"timestamp":"1629798921.036759138","source":"terraform@/home/vcap/tmp/gsb1178090049","message":"terraform@/home/vcap/tmp/gsb1178090049.terraform execution failed","log_level":2,"data":{"correlation-id":"639a9561-a553-4864-6526-a79ce57c1830","error":"exit status 1","errors":"\nError: expected max_size_gb to be in the range (1 - 4096), got 0\n\n  on main.tf line 2, in resource \"azurerm_mssql_database\" \"primary_db\":\n   2: resource \"azurerm_mssql_database\" \"primary_db\" {\n\n\n","request-id":"a442c2f9-5a51-4428-befb-ab5c9cff0ed5"}}
...
```

And with databases that are 1 GB and larger I am currently seeing successful subsume with no destructive behaviour.

From the timeout messages that you've posted above, it looks like the connection between the Azure Terraform Provider and Azure is broken. My guess is that the provider sent an API request to Azure and never got a response.

I'm wondering whether the Azure Monitor Activity Log shows more information? Maybe you have looked at this already. But it would be interesting to know why Azure thinks that the database has been removed. Did it get a specific "delete" command, or was there an error?

To summarise where we are, I think you are seeing destructive behaviour when subsuming for DB FOGs that are smaller than 1 GB, but also for ones that are larger than 1 GB. And this is only seen when the database is nearly (around 96%) full. I haven't yet managed to reproduce this, so it would be good to consider potential differences between the environments that we are using. Are you using any custom plans, or have and adjustments or additions to the Terraform? Or are there any other differences that you're aware of that might explain why we see different results?

**Comment by kelvin-j-li** [comment]
>> To summarise where we are, I think you are seeing destructive behaviour when subsuming for DB FOGs that are smaller than 1 GB, but also for ones that are larger than 1 GB. And this is only seen when the database is nearly (around 96%) full. 

yes, that's right. My initial issue (risk item 2) was with a db (2G) near capacity in UAT app -- it failed twice, only succeeded after increasing the db storage size. When I tested it in my sample app with 0.2.7+1.0.0-rc.40, I used a small db (100MB) size and triggered a different destructive case as you mentioned. I will re-test it again with db size 2G. 

>> I'm wondering whether the Azure Monitor Activity Log shows more information? Maybe you have looked at this already. But it would be interesting to know why Azure thinks that the database has been removed. Did it get a specific "delete" command, or was there an error?

No, I didn't check Azure Monitor Actvity Log when it (risk item 1 - Future#WaitforCompletion). I will check it when I test the new version (0.2.8) on Friday for a db with high DTU usage.
